Feature/unit tests

aflegel opened the pull request.

alexanderboutinflegelphl committed:
updated the editor config with new warnings and errors
refactored to use expressions
refactored for more linq expressions over for loops

alexanderboutinflegelphl committed:
Added some unit tests
exposed some functions for the tester.

alexanderboutinflegelphl committed:
Code cleanup on seeded data
Simplified logic to use inline expressions

alexanderboutinflegelphl committed:
Updated the generators to be static extensions
reinstated some default constructors to allow linq to work
Updated tests to include more complicated pools of dice

alexanderboutinflegelphl committed:
Cleaned up the unit tests.

alexanderboutinflegelphl committed:
Moved the data seed into the generator project
Moved the positive and negative die type lists to the framework

aflegel committed:
Cleanup
Fixed the intersect function for GetPoolId

alexanderboutinflegelphl committed:
Migrated the generator project into the data framework as the seed data

alexanderboutinflegelphl committed:
Cleaned up the seeding functions
moved the common functions into different objects
modified the search to use the pool name instead of searching by children

aflegel committed:
Optimizing the seed methods
removing dead extensions

alexanderboutinflegelphl committed:
Cleaned up the split function to use linq to take and skip
merged the escape function for the recursion call into the recursion fucntion
Explode the dice during calculations as splitting and grouping was wasteful.

alexanderboutinflegelphl committed:
Cleaned up the statistics generation
Removed the dead equatable code.

alexanderboutinflegelphl committed:
Modified the context load configuration externally
Purged IEquatable code that was dead.
WIP on API tests
WIP on database Tests

aflegel committed:
Fixed the seeding of the api testing context.
Renamed functions for clarity

alexanderboutinflegelphl committed:
Corrected API Tests to not double seed the db
Added roll results controller tests
improved the gethashcode functions
removed the int mapping from the symbol enum
updated the editorconfig

alexanderboutinflegelphl committed:
Moved some generation code out of the program file and into a new seed file
named the seed files to match their classes.

aflegel committed:
Moved the context accessors into the framework project
removed dead extensions
split extensions into extension files
cleaned up the split and topool functions
renamed the PoolText function to ToString

aflegel committed:
Removed the old json ignore tags

aflegel committed:
Nuked unused imports.

aflegel committed:
Converted the split functions to use tuples.
Optimized the recursion function

alexanderboutinflegelphl committed:
Started using Value Tuples <3
Moved some of the pool seed funtions into PoolResultSeed
Moved more db accessors to the accessors extension

alexanderboutinflegelphl committed:
Documentation and better function names.
Added some fail tests

alexanderboutinflegelphl committed:
Added some configuration settings
Modified the Combination Query to include the die (I don't know how this was working before)

aflegel committed:
WIP on making the seed async.
Made the seeding program a little more functional

alexanderboutinflegelphl committed:
WIP on model cleanup

alexanderboutinflegelphl committed:
Merge WIPS

alexanderboutinflegelphl committed:
Upgraded tests to more accurately reflect the seeding methods
Cleaned up unused accessors
modified the db seed to be more isolated.
The base generation should be thread safe-ish now.

ChrisDusyk commented on DataFramework/Context/ProbabilityContext.Accessors.cs:
```diff
@@ -0,0 +1,81 @@
+			.Select(s => s.PoolId)).Aggregate((result, next) => result.Intersect(next)).FirstOrDefault();
+
+		/// <summary>
+		/// Trys to find the pool id from a given name
```
`Trys` spelling mistake

ChrisDusyk commented on DataFramework/Context/ProbabilityContext.Accessors.cs:
```diff
@@ -0,0 +1,81 @@
+		public static int? GetPoolIdByName(this ProbabilityContext context, string poolName) => context.Pools.Where(w => w.Name == poolName).Select(s => s.PoolId).FirstOrDefault();
+
+		/// <summary>
+		/// Trys to get the pool ids split by positive and negative dice
```
`Trys` again 🤔

ChrisDusyk commented on DataFramework/Models/PoolResult.cs:
```diff
@@ -44,5 +26,7 @@ public PoolResult(List<PoolResultSymbol> PoolResultSymbols)
 		/// <returns></returns>
 		public int CountMatchingKeys(Symbol key) => PoolResultSymbols.Where(a => a.Symbol == key).Sum(s => s.Quantity);
+
+		public override int GetHashCode() => PoolResultSymbols.Select(s => s.GetHashCode()).Aggregate(1, (a, b) => unchecked(a * b));
```
This is clever

ChrisDusyk commented on Visualizer.Tests/APITests.cs:
```diff
@@ -0,0 +1,112 @@
+			Assert.True(15 == result.PositiveRolls.Results.Count(), $"Statistics count incorrect.  Count was {result.PositiveRolls.Results.Count()}");
+			Assert.True(15 == result.NegativeRolls.Results.Count(), $"Statistics count incorrect.  Count was {result.PositiveRolls.Results.Count()}");
+
+			controller.Dispose();
```
Is there a specific reason to `Dispose` of these controllers explicitly, vs letting GC handle them?

aflegel replied:
It complained at me actually.  

I think this is normally hooked up as part of the services but running it in a test it doesn't get the normal treatment.

ChrisDusyk commented on Visualizer.Tests/SeedingTests.cs:
```diff
@@ -0,0 +1,82 @@
+		public void DieComparisonGeneratorAdvanced()
+		{
+			var pools = DiceSeed.SeedDice().ProcessPools((Enumerable.Range(1, 2), Enumerable.Range(1, 1), Enumerable.Range(0, 1)), (Enumerable.Range(1, 2), Enumerable.Range(1, 1), Enumerable.Range(0, 1))).CrossProduct();
+			//var pool = pools.First();
```
Cleanup

ChrisDusyk commented on Visualizer/Models/DieViewModel.cs:
```diff
@@ -1,9 +1,19 @@
 		public int Quantity { get; set; }
 	}
+
+	public static class DieViewModelExtensions
```
Might want to split this into its own file?

ChrisDusyk reviewed (approved):
This was a lot of review! So I think you've done a good job refactoring, though it's kind of tough to review a refactor that well.

Something you might want to look into for your tests is [`FluentAssertions`](https://fluentassertions.com/). It'll make them nicer to read.

aflegel committed:
Included FluentAssertions to clean up tests
Reworked tests to use a Database Fixture
Separated the controllers into separate test files
Package updates

aflegel committed:
Test Cleamup

ChrisDusyk reviewed (approved):
I think those were some good changes, especially for the tests

alexanderboutinflegelphl committed:
Updated the pool combination seed to take a tuple and return a PoolCombination object
modified the readonly lists to use Immutable lists
